x-pack/plugins/fleet/server/services/spaces/enable_space_awareness.ts

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { SavedObjectsClientContract } from '@kbn/core-saved-objects-api-server';
+
+import { LEGACY_AGENT_POLICY_SAVED_OBJECT_TYPE } from '../../../common';
+import {
+  LEGACY_PACKAGE_POLICY_SAVED_OBJECT_TYPE,
+  PACKAGE_POLICY_SAVED_OBJECT_TYPE,
+} from '../../../common/constants';
+
+import { appContextService } from '..';
+import { AGENT_POLICY_SAVED_OBJECT_TYPE } from '../../constants';
+import { getSettingsOrUndefined, saveSettings } from '../settings';
+import { FleetError } from '../../errors';
+
+const PENDING_TIMEOUT = 60 * 60 * 1000;
+
+export async function enableSpaceAwarenessMigration() {
+  const soClient = appContextService.getInternalUserSOClientWithoutSpaceExtension();
+  const logger = appContextService.getLogger();
+  try {
+    const existingSettings = await getSettingsOrUndefined(soClient);
+    if (existingSettings?.use_space_awareness_migration_status === 'success') {
+      return;
+    }
+
+    if (
+      existingSettings?.use_space_awareness_migration_started_at &&
+      new Date(existingSettings?.use_space_awareness_migration_started_at).getTime() >
+        Date.now() - PENDING_TIMEOUT
+    ) {
+      throw new FleetError('Migration is pending.');
+    }
+
+    logger.info('Starting Fleet space awareness migration');
+
+    await saveSettings(
+      soClient,
+      {
+        use_space_awareness_migration_status: 'pending',
+        use_space_awareness_migration_started_at: new Date().toISOString(),
+      },
+      {
+        version: existingSettings?.version,
+      }
+    );
+
+    // Migration
+    // Agent Policy
+    await batchMigration(
+      soClient,
+      LEGACY_AGENT_POLICY_SAVED_OBJECT_TYPE,
+      AGENT_POLICY_SAVED_OBJECT_TYPE
+    );
+    await batchMigration(
+      soClient,
+      LEGACY_PACKAGE_POLICY_SAVED_OBJECT_TYPE,
+      PACKAGE_POLICY_SAVED_OBJECT_TYPE
+    );
+
+    // Update Settings SO
+    await saveSettings(soClient, {
+      use_space_awareness_migration_status: 'success',
+      use_space_awareness_migration_started_at: null,
+    });
```

Comment:
Concurrency control _should_ be happening automatically, but it may make sense to, inside the `saveSettings` function add the SO `version` to the `soClient.update` call. This means that if this endpoint gets called twice in short succession one of the other calls will fail early.

You get the SO version when `find`ing the SO object. So you would have:

```ts
const { saved_objects: [object] } = savedObjects.find(...);

const res = savedObjects.update(type, id, data, { version: object.version })
```

Reply:
there are other code path that could save the settings, I think here we should not fail if such a request happen before as we are setting the final state of the migration. 

The code path where we should fail (and we already pass the version is when we are creating the lock)